mmv1/third_party/terraform/utils/cloudrun_polling.go

```diff
@@ -26,37 +26,63 @@ type KnativeStatus struct {
 		SelfLink  string
 	}
 	Status struct {
-		Conditions []Condition
+		Conditions         []Condition
+		ObservedGeneration float64
 	}
 }
 
-func PollCheckKnativeStatus(resp map[string]interface{}, respErr error) PollResult {
-	if respErr != nil {
-		return ErrorPollResult(respErr)
+func getGeneration(res map[string]interface{}) (int, error) {
+	metadata, ok := res["metadata"]
+	if !ok {
+		return 0, fmt.Errorf("Unable to find knative metadata")
 	}
-	s := KnativeStatus{}
-	if err := Convert(resp, &s); err != nil {
-		return ErrorPollResult(errwrap.Wrapf("unable to get KnativeStatus: {{err}}", err))
+	m, ok := metadata.(map[string]interface{})
+	if !ok {
+		return 0, fmt.Errorf("Unable to find generation in knative metadata")
 	}
+	gen, ok := m["generation"]
+	if !ok {
+		return 0, fmt.Errorf("Unable to find generation in knative metadata")
+	}
+	return int(gen.(float64)), nil
+}
 
-	for _, condition := range s.Status.Conditions {
-		if condition.Type == readyStatusType {
-			log.Printf("[DEBUG] checking KnativeStatus Ready condition %s: %s", condition.Status, condition.Message)
-			switch condition.Status {
-			case "True":
-				// Resource is ready
-				return SuccessPollResult()
-			case "Unknown":
-				// DomainMapping can enter a 'terminal' state where "Ready" status is "Unknown"
-				// but the resource is waiting for external verification of DNS records.
-				if condition.Reason == pendingCertificateReason {
+func PollCheckKnativeStatusFunc(res map[string]interface{}) func(resp map[string]interface{}, respErr error) PollResult {
+	return func(resp map[string]interface{}, respErr error) PollResult {
+		if respErr != nil {
+			return ErrorPollResult(respErr)
+		}
+		s := KnativeStatus{}
+		if err := Convert(resp, &s); err != nil {
+			return ErrorPollResult(errwrap.Wrapf("unable to get KnativeStatus: {{err}}", err))
+		}
+
+		gen, err := getGeneration(res)
+		if err != nil {
+			return ErrorPollResult(errwrap.Wrapf("unable to find Knative generation: {{err}}", err))
```

Comment:
I don't think this would compile? This function returns a function that takes a response + response error and returns a PollResult. Your change would make this function return a PollResult or a function that returns a poll result

I nested the generation extraction within the so that it would be able to handle errors via the normal ErrorPollResult if generation extraction fails for some reason. Due to the code generation & how this is called it is difficult to extract the generation prior to calling this function

Reply:
ah yeah that makes sense